File: backend/aircraft_overrides.py
"""Applies user-supplied corrections on top of scenery_data.scan_aircraft_and_liveries'
best-effort classification (manifest content_type for aircraft-vs-livery,
base_container matching for a livery's parent aircraft) - both are
heuristics and can be wrong (a payware "livery" manifested as AIRCRAFT, or
a livery whose base_container doesn't resolve), so the user can override
either per package. Pure logic, no UI/Tk dependency, same spirit as
scenery_data.py itself (which stays unmodified).
"""


def apply_overrides(aircraft, liveries, type_overrides, parent_overrides, dismissed_suggestions=None):
    """type_overrides: {folder_name: "aircraft"|"livery"} - reclassifies a
    package regardless of what its own manifest said.
    parent_overrides: {folder_name: parent_folder_name} where "" means
    "explicitly unassigned" and an absent key means "use the heuristic
    (base_container) result".
    dismissed_suggestions: folder_names the user has dismissed the
    "suspected_livery" hint for (see scenery_data.py) - stops the hint from
    showing again without forcing any actual reclassification.

    Returns (aircraft, liveries) - new lists, inputs are not mutated.
    """
    dismissed_suggestions = dismissed_suggestions or set()
    combined = {r["folder_name"]: r for r in aircraft + liveries}
    original_aircraft_folders = {r["folder_name"] for r in aircraft}

    def is_aircraft(folder_name):
        override = type_overrides.get(folder_name)
        if override == "aircraft":
            return True
        if override == "livery":
            return False
        return folder_name in original_aircraft_folders

    final_aircraft_folders = {fn for fn in combined if is_aircraft(fn)}

    result_aircraft = []
    result_liveries = []
    for folder_name, record in combined.items():
        record = dict(record)
        if is_aircraft(folder_name):
            record.pop("parent_aircraft", None)
            # Once the user has made any explicit decision about this
            # package (an override either way, or an explicit dismissal),
            # the suggestion has done its job - stop nagging about it.
            if folder_name in type_overrides or folder_name in dismissed_suggestions:
                record["suspected_livery"] = False
            result_aircraft.append(record)
        else:
            if folder_name in parent_overrides:
                parent = parent_overrides[folder_name] or None
            else:
                parent = record.get("parent_aircraft")
            if parent not in final_aircraft_folders:
                parent = None
            record["parent_aircraft"] = parent
            result_liveries.append(record)

    def _sort_key(r):
        return (r["developer"] or "￿", r["display_name"].lower())

    result_aircraft.sort(key=_sort_key)
    result_liveries.sort(key=_sort_key)
    return result_aircraft, result_liveries

File: dotnet/src/FlightOpsHub.Core/AircraftOverrides.cs
using System.Text.Json.Nodes;

namespace FlightOpsHub.Core;

/// <summary>
/// Port of backend/aircraft_overrides.py - applies user-supplied
/// corrections on top of AircraftScanner's best-effort classification
/// (manifest content_type for aircraft-vs-livery, base_container matching
/// for a livery's parent aircraft) - both are heuristics and can be wrong,
/// so the user can override either per package.
/// </summary>
public static class AircraftOverrides
{
    /// <summary>
    /// typeOverrides: {folderName: "aircraft"|"livery"} - reclassifies a
    /// package regardless of what its own manifest said.
    /// parentOverrides: {folderName: parentFolderName} where "" means
    /// "explicitly unassigned" and an absent key means "use the heuristic
    /// (base_container) result".
    /// dismissedSuggestions: folder names the user has dismissed the
    /// "suspected_livery" hint for - stops the hint from showing again
    /// without forcing any actual reclassification.
    ///
    /// Returns new lists; inputs are not mutated.
    /// </summary>
    public static (List<JsonObject> Aircraft, List<JsonObject> Liveries) ApplyOverrides(
        IReadOnlyList<JsonObject> aircraft,
        IReadOnlyList<JsonObject> liveries,
        IReadOnlyDictionary<string, string> typeOverrides,
        IReadOnlyDictionary<string, string> parentOverrides,
        IReadOnlySet<string>? dismissedSuggestions = null)
    {
        dismissedSuggestions ??= new HashSet<string>();

        var combined = new Dictionary<string, JsonObject>();
        foreach (var r in aircraft.Concat(liveries))
        {
            combined[r["folder_name"]!.GetValue<string>()] = r;
        }
        var originalAircraftFolders = aircraft.Select(r => r["folder_name"]!.GetValue<string>()).ToHashSet();

        bool IsAircraft(string folderName)
        {
            if (typeOverrides.TryGetValue(folderName, out var over))
            {
                if (over == "aircraft") return true;
                if (over == "livery") return false;
            }
            return originalAircraftFolders.Contains(folderName);
        }

        var finalAircraftFolders = combined.Keys.Where(IsAircraft).ToHashSet();

        var resultAircraft = new List<JsonObject>();
        var resultLiveries = new List<JsonObject>();

        foreach (var (folderName, original) in combined)
        {
            var record = (JsonObject)original.DeepClone();
            if (IsAircraft(folderName))
            {
                record.Remove("parent_aircraft");
                // Once the user has made any explicit decision about this
                // package (an override either way, or an explicit
                // dismissal), the suggestion has done its job.
                if (typeOverrides.ContainsKey(folderName) || dismissedSuggestions.Contains(folderName))
                {
                    record["suspected_livery"] = false;
                }
                resultAircraft.Add(record);
            }
            else
            {
                string? parent;
                if (parentOverrides.TryGetValue(folderName, out var overriddenParent))
                {
                    parent = string.IsNullOrEmpty(overriddenParent) ? null : overriddenParent;
                }
                else
                {
                    parent = record["parent_aircraft"]?.GetValue<string>();
                }
                if (parent is null || !finalAircraftFolders.Contains(parent))
                {
                    parent = null;
                }
                record["parent_aircraft"] = parent;
                resultLiveries.Add(record);
            }
        }

        static int SortCompare(JsonObject a, JsonObject b)
        {
            var devA = a["developer"]?.GetValue<string>();
            var devB = b["developer"]?.GetValue<string>();
            var devCompare = string.CompareOrdinal(string.IsNullOrEmpty(devA) ? "￿" : devA, string.IsNullOrEmpty(devB) ? "￿" : devB);
            if (devCompare != 0) return devCompare;
            return string.CompareOrdinal(
                a["display_name"]!.GetValue<string>().ToLowerInvariant(),
                b["display_name"]!.GetValue<string>().ToLowerInvariant());
        }

        resultAircraft.Sort(SortCompare);
        resultLiveries.Sort(SortCompare);
        return (resultAircraft, resultLiveries);
    }
}
